mozetl/maudau/maudau.py

```diff
@@ -0,0 +1,214 @@
+        for row in reader:
+            if row['day'] >= since:
+                break
+            carryover.append(row)
+    return since, carryover
+
+
+def get_last_rollup(transferer):
+    '''
+    Always chose the latest rollup, even if start_date is old.
+    '''
+    basename = MAUDAU_ROLLUP_BASENAME
+    key = os.path.join(STORAGE_SUB_DIR, basename)
+    try:
+        transferer.download_file(
+            STORAGE_BUCKET,
+            key,
+            basename)
+    except botocore.exceptions.ClientError as e:
+        # If the file wasn't there, that's ok. Otherwise, abort!
+        if e.response['Error']['Code'] != "404":
+            raise e
+        else:
+            msg = "Did not find an existing file at '{}'".format(basename)
+            logger.info(msg)
+        return None
+    else:
+        return basename
+
+
+def generate_counts(frame, since, until=None):
+    cols = [frame.client_id.alias('clientId'),
+            'subsession_start_date', 'submission_date_s3']
+    narrow = frame.select(cols)
+    updates = []
+    generated = format_as_submission_date(today.isoformat())
+    start = DT.datetime.strptime(since, '%Y%m%d').date()
+    until = (DT.datetime.strptime(until, '%Y%m%d').date() if until
+             else today)
+    while start < until:
+        dau = count_active_users(narrow, start, 0)
+        mau = count_active_users(narrow, start, MONTH)
+        day = format_as_submission_date(start.isoformat())
+        d = {'day': day, 'dau': dau, 'mau': mau, 'generated_on': generated}
+        updates.append(d)
+        start += DT.timedelta(1)
+    return updates
+
+
+def write_locally(results):
+    '''
+    :results [{'day': '%Y%m%d', 'dau': <int>,
+              'mau': <int>, 'generated_on': '%Y%m%d'}, ...]
+    '''
+    publication_date = format_as_submission_date(today.isoformat())
+    basename = MAUDAU_SNAPSHOT_TEMPLATE.format(publication_date)
+    cols = ["day", "dau", "mau", "generated_on"]
+    with open(basename, 'w') as f:
+        writer = csv.DictWriter(f, cols)
+        writer.writeheader()
+        writer.writerows(results)
+    return basename
+
+
+def publish_to_s3(transferer, basename):
+    '''
+    Write the file twice to storage (once dated, once as "latest")
+    and once to the production dashboard.
+    '''
+    dated_key = os.path.join(STORAGE_SUB_DIR, basename)
+    transferer.upload_file(basename, STORAGE_BUCKET, dated_key)
+    latest_key = os.path.join(STORAGE_SUB_DIR, MAUDAU_ROLLUP_BASENAME)
+    transferer.upload_file(basename, STORAGE_BUCKET, latest_key)
+    if DEVELOPMENT:
+        return
+    transferer.upload_file(basename, DASHBOARD_BUCKET, DASHBOARD_KEY)
+
+
+@click.command()
+@click.option('--input_bucket',
+              default='telemetry-parquet',
+              help='Bucket of the input dataset')
+@click.option('--input_prefix',
+              default='main_summary/v4',
+              help='Prefix of the input dataset')
+def main(input_bucket, input_prefix):
```

Comment:
Having the output bucket and prefix would be useful for testing on airflow -- the staging instance can specify a testing bucket so it doesn't affect the production files.